clang/lib/Sema/SemaInit.cpp

```diff
@@ -7575,15 +7577,27 @@ static void visitLifetimeBoundArguments(IndirectLocalPath &Path, Expr *Call,
     Path.pop_back();
   };
 
-  if (ObjectArg && implicitObjectParamIsLifetimeBound(Callee))
-    VisitLifetimeBoundArg(Callee, ObjectArg);
-
   bool CheckCoroCall = false;
   if (const auto *RD = Callee->getReturnType()->getAsRecordDecl()) {
     CheckCoroCall = RD->hasAttr<CoroLifetimeBoundAttr>() &&
                     RD->hasAttr<CoroReturnTypeAttr>() &&
                     !Callee->hasAttr<CoroDisableLifetimeBoundAttr>();
   }
+
+  if (ObjectArg) {
+    bool CheckCoroObjArg = CheckCoroCall;
+    // Ignore `__promise.get_return_object()` as it is not lifetimebound.
+    if (CheckCoroObjArg && Callee->getDeclName().isIdentifier() &&
+        Callee->getName() == "get_return_object")
+      CheckCoroObjArg = false;
+    // Coroutine lambda objects with empty capture list are not lifetimebound.
+    if (auto *LE = dyn_cast<LambdaExpr>(ObjectArg->IgnoreImplicit());
+        LE && LE->captures().empty())
+      CheckCoroObjArg = false;
```

Comment:
I am a little bit surprised that we need to handle this specially. Is this an optimization? Or it is necessary and why?

Reply:
This is a false positive involving lambda with no capture list. The following code does not have lifetime issue but will be warned otherwise
```
auto x = []() -> coro<int> { co_return 0; }();
```

Reply:
I am curious why it is. Have you taken a look on that?

Reply:
I can explain. The lambda call operator is a coroutine and receives a lambda object as an implicit `this` param.
This lambda object would not be used or read in the coroutine body (as there is no capture list). But this analysis would still consider the object argument to be lifetime bound to the return object and would flag it the return object outlives the lambda object (which is true in the mentioned example).

Coro_lifetimebound, by design, has false positives because of it strictness. It considers all args to be lifetimebound but it might not be true in practice, for eg. if the arg is not used by the coro or used before first suspend.
Lambda with no capture list allows us to recognize one such instance where it can be proven to be a false positive always.

Reply:
Thanks for the explanation. Got it. Then it looks an optimization for lambdas for me.

Reply:
It may not result in runtime errors (even in sanitizers), because this is never accessed, but I wonder if it is actually legal or UB in the standard?

AFAIK, calling a method if the object argument is `null` is UB, even if `this` is never accessed inside the method itself (citation needed, though, I looked at section of class member access and function call, but couldn't find a decisive wording there).

So, a coroutine frame ends up keeping a pointer to a lambda that goes out of scope. After initial suspension and on resume, it will probably **still call a method on that same pointer** per C++ standard.
Which makes me skew in the direction of warning for these cases two (but we need some proof from the standard gurus here)

Reply:
There are simple workarounds to silence these warnings in those cases:
- converting to a function pointer before call `(+[](){ ... co_await foo(); ... })()`,
- declaring a lambda as static  `[]() static { ... co_await foo(); ...}()` (starting with C++23).

Reply:
> It may not result in runtime errors (even in sanitizers), because this is never accessed, but I wonder if it is actually legal or UB in the standard?

I feel it is literally undefined. Since the spec doesn't say a lot about resumption/suspension about coroutines. Also the description of http://eel.is/c++draft/dcl.fct.def.coroutine#note-3 is vague too. It says it is `likely` to be an undefined behavior.

> Which makes me skew in the direction of warning for these cases two

I didn't get the logic here. Why it is not good to warn the undefined things?

Reply:
> I didn't get the logic here. Why it is not good to warn the undefined things?

Sorry, I wasn't clear here. I meant that if it is UB, we should warn for those cases too and remove this code path.

> I feel it is literally undefined. Since the spec doesn't say a lot about resumption/suspension about coroutines. Also the description of http://eel.is/c++draft/dcl.fct.def.coroutine#note-3 is vague too. It says it is likely to be an undefined behavior.

Yeah, this looks suspicious. At the same time, it's very similar to `delete this`, which (IIUC) is not UB as long as people are careful to not access `this` after doing it.

I have actually asked the CWG, but got no reply so far.

Reply:
> > I didn't get the logic here. Why it is not good to warn the undefined things?
> 
> Sorry, I wasn't clear here. I meant that if it is UB, we should warn for those cases too and remove this code path.
> 
> > I feel it is literally undefined. Since the spec doesn't say a lot about resumption/suspension about coroutines. Also the description of http://eel.is/c++draft/dcl.fct.def.coroutine#note-3 is vague too. It says it is likely to be an undefined behavior.
> 
> Yeah, this looks suspicious. At the same time, it's very similar to `delete this`, which (IIUC) is not UB as long as people are careful to not access `this` after doing it.
> 
> I have actually asked the CWG, but got no reply so far.

Got it. Then it might not be a UB (we didn't access `this` in the empty lambda) if we're comparing this case. While I don't object wait for response from CWG, I feel we can continue here. On the one hand, we're talking about the behaviors with an extension (`[[clang::coro_lifetime_bound]]`) which can be visited as a dialect. On the other hand, it is somewhat not user friendly to emit warnings in such cases.

Reply:
I would be in favor of continuing here (without warning for empty-capture lambdas). We can revisit it if the CWG clarifies this as a UB.

Reply:
The only downside of continuing as-is is that people who start using this warning might have to rewrite their code *if CWG responds that it is UB*.

Even though I don't expect to get an answer it's UB because there is a conversion to a function pointer on the same lambdas that's supposed to work in that case. The most natural implementation is to share the code between wherever call operators forwards to and where function pointers points at. So both should either work or not.

I guess this leads me to think that the approach of not showing the warning for empty lambdas is ok as the answer that it's UB is highly unlikely (IMO).

So LG.